core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -25,34 +37,295 @@ type GCSFileTransfer struct {
+	if err != nil {
+		return formatDownloadError(
+			fmt.Sprintf("error copying file %s", localPath),
+			err,
+		)
+	}
+
+	return nil
+}
+
+// getObjectAndAttrs returns the object handle and attributes if the object
+// exists and corresponds with a file.
+//
+// If the object corresponds to a directory, we return the ErrObjectIsDirectory
+// error to skip downloading without failing.
+func (ft *GCSFileTransfer) getObjectAndAttrs(
+	bucket *storage.BucketHandle,
+	task *ReferenceArtifactDownloadTask,
+	key string,
+) (*storage.ObjectHandle, *storage.ObjectAttrs, error) {
+	if strings.HasSuffix(key, "/") {
+		return nil, nil, ErrObjectIsDirectory
+	}
+	object, err := ft.getObject(bucket, task, key)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	attrs, err := object.Attrs(ft.ctx)
+	// if object doesn't have an extension and has size 0, it might be a
+	// directory as we cut off the ending "/" in the manifest entry.
+	if errors.Is(err, storage.ErrObjectNotExist) &&
+		(path.Ext(key) == "" && task.Size == 0) {
+		key := key + "/"
+		object, err := ft.getObject(bucket, task, key)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// if the new key corresponds with an existing directory, it should
+		// return the attrs without an error
+		_, err = object.Attrs(ft.ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+		return nil, nil, ErrObjectIsDirectory
+	} else if err != nil {
+		return nil, nil, err
+	}
+	return object, attrs, nil
+}
+
+// getObject returns the object handle given a bucket, key, and task object
+// with a possible versionId.
+func (ft *GCSFileTransfer) getObject(
+	bucket *storage.BucketHandle,
+	task *ReferenceArtifactDownloadTask,
+	key string,
+) (*storage.ObjectHandle, error) {
+	object := bucket.Object(key)
+	if task.VersionId != nil {
+		versionId, err := safeConvertToInt64(task.VersionId)
+		if err != nil {
+			err := fmt.Errorf("failed to convert VersionId: %v", err)
+			return nil, err
+		}
+		object = object.Generation(versionId)
+	}
+	return object, nil
+}
+
+// parseReference parses the reference path and returns the bucket name and
+// object name.
+func parseReference(reference string) (string, string, error) {
+	uriParts, err := url.Parse(reference)
+	if err != nil {
+		return "", "", err
+	}
+	if uriParts.Scheme != "gs" {
+		err := fmt.Errorf("invalid gsutil URI %s", reference)
+		return "", "", err
+	}
+	bucketName := uriParts.Host
+	objectName := strings.TrimPrefix(uriParts.Path, "/")
+	return bucketName, objectName, nil
+}
+
+// getDownloadFilePath returns the file path to download the file to.
+//
+// We expect the base file path to already have the prefix appended to it, so
+// we remove any part of the object name that is already contained in the base
+// file path before appending it.
+func getDownloadFilePath(objectName string, prefix string, baseFilePath string) string {
+	ext, _ := strings.CutPrefix(objectName, prefix)
+	localPath := baseFilePath + ext
+	return localPath
+}
```

Comment:
moving the logic inline makes sense to me. I don't know if we want to remove task altogether from `downloadFiles` since we use it quite a bit (for size, `shouldCheckDigest`, the digest itself, etc), so we'd have to pass in a lot more variables, but I'll pass in root object name to skip parsing again